src/stdlib_experimental_quadrature_trapz.fypp

```diff
@@ -0,0 +1,86 @@
+#:include "common.fypp"
+
+submodule (stdlib_experimental_quadrature) stdlib_experimental_quadrature_trapz
+
+    implicit none
+
+contains
+
+    #:for KIND in REAL_KINDS
+    
+    pure module function trapz_dx_${KIND}$(y, dx) result(integral)
+        real(${KIND}$), dimension(:), intent(in) :: y
+        real(${KIND}$), intent(in) :: dx
+        real(${KIND}$) :: integral
+
+        integer :: n
+
+        n = size(y)
+        
+        select case (n)
+        case (0:1)
+            integral = 0.0_${KIND}$
+        case (2)
+            integral = 0.5_${KIND}$*dx*(y(1) + y(2))
+        case default
+            integral = dx*(sum(y(2:n-1)) + 0.5_${KIND}$*(y(1) + y(n)))
+        end select
+    end function trapz_dx_${KIND}$
+    
+    #:endfor
+    #:for KIND in REAL_KINDS
+    
+    pure module function trapz_x_${KIND}$(y, x) result(integral)
+        real(${KIND}$), dimension(:), intent(in) :: y
+        real(${KIND}$), dimension(size(y)), intent(in) :: x
+        real(${KIND}$) :: integral
+
+        integer :: i
+        integer :: n
+
+        n = size(y)
+
+        select case (n)
+        case (0:1)
+            integral = 0.0_${KIND}$
+        case (2)
+            integral = 0.5_${KIND}$*(x(2) - x(1))*(y(1) + y(2))
+        case default
+            integral = 0.0_${KIND}$
+            do i = 2, n
+                integral = integral + (x(i) - x(i-1))*(y(i) + y(i-1))
+            end do
+            integral = 0.5_${KIND}$*integral
+        end select
+    end function trapz_x_${KIND}$
+
+    #:endfor
+    #:for KIND in REAL_KINDS
+
+    pure module function trapz_weights_${KIND}$(x) result(w)
+        real(${KIND}$), dimension(:), intent(in) :: x
+        real(${KIND}$), dimension(size(x)) :: w
+
+        integer :: i
+        integer :: n
+
+        n = size(x)
+
+        select case (n)
+        case (0)
+            ! no action needed
```

Comment:
I agree that "no action" is needed. But should the function not return somethin for `w` (e.g. NaN, a huge number, 0,...)? Currently `w` is undefined.

The other functions return 0 for `case (0)`.

Reply:
For `trapz`, zero-size inputs behave the same as summing a zero-size array, producing zero. For `trapz_weights`, you want the result to behave such that `sum(w*x)` also produces zero. 

There are also practical challenges to returning anything besides a zero-size array. It's OK that it's value is undefined. That's just how zero-size arrays go.